backend-transit/src/main/java/eu/socialedge/hermes/backend/transit/domain/GoogleMapsShapeFactory.java

```diff
@@ -64,45 +62,79 @@ public Shape create(List<Location> waypoints) {
+            val wpChunk = Arrays.copyOfRange(waypoints, wpChunkFromIndex, wpChunkToIndex); // 'to' is exclusive
+            val pathDistances = calculatePathDistances(wpChunk);
 
-            for(int j = 0; j < wpChunk.length - 1; j++) {
-                val origDestDistanceMatrix = wpChunkDistanceMatrix.rows[j].elements[j];
+            for (val origDestDistance : pathDistances.entrySet()) {
+                val originDestLatLngPair = origDestDistance.getKey();
+                val destLatLng = originDestLatLngPair.getValue();
 
-                if (origDestDistanceMatrix.status != DistanceMatrixElementStatus.OK)
-                    throw new ShapeFactoryException("Couldn't recognize location " + waypoints[j + 1]);
+                val originDestDistance = origDestDistance.getValue();
+                distTraveled = distTraveled.add(originDestDistance);
 
-                val origDestDistance = origDestDistanceMatrix.distance.inMeters;
-                distTraveled = distTraveled.add(Quantities.getQuantity(origDestDistance, METRE));
-
-                waypointDistTraveled.add(new Pair<>(waypoints[j + 1], distTraveled));
+                pathDistanceTraveled.put(destLatLng, distTraveled);
             }
         }
 
-        return waypointDistTraveled;
+        return pathDistanceTraveled;
     }
 
-    private DistanceMatrix calculateDistanceMatrix(LatLng[] waypoints) {
+    private Map<Pair<LatLng, LatLng>, Quantity<Length>> calculatePathDistances(LatLng[] waypoints) {
+        if (waypoints.length == 0)
+            return Collections.emptyMap();
+
+        val pathDistances = new LinkedHashMap<Pair<LatLng, LatLng>, Quantity<Length>>();
+
         val latLngOrigins = Arrays.copyOfRange(waypoints, 0, waypoints.length - 1);
         val latLngDestinations = Arrays.copyOfRange(waypoints, 1, waypoints.length);
+        val pathDistanceMatrix = calculateDistanceMatrix(latLngOrigins, latLngDestinations);
+
+        for (int i = 0; i < pathDistanceMatrix.rows.length; i++) {
+            val originDistanceMatrix = pathDistanceMatrix.rows[i];
+            val originDiagonalDestDistanceElement = originDistanceMatrix.elements[i];
+
+            if (originDiagonalDestDistanceElement.status != DistanceMatrixElementStatus.OK)
+                throw new ShapeFactoryException("Couldn't recognize location " + latLngOrigins[i]);
+
+            val originDestDistanceRaw = originDiagonalDestDistanceElement.distance.inMeters;
+
+            val originLatLng = latLngOrigins[i];
+            val destLatLng = latLngDestinations[i];
+            val originDestDistance = Quantities.getQuantity(originDestDistanceRaw, METRE);
+
+            pathDistances.put(Pair.of(originLatLng, destLatLng), originDestDistance);
+        }
+
+        return pathDistances;
+    }
+
+    private DistanceMatrix calculateDistanceMatrix(LatLng[] origins, LatLng[] destinations) {
+        if (origins.length >= LOCATIONS_LIMIT)
```

Comment:
I'm not sure the thought "if it's private method I should assume it will receive valid data". Mistake in a code that calls a private method is an exceptional situation too, like ArrayOutOfTheBoundException or ArithmeticException. It's better to fail fast here for invalid request instead of wasting API calls imho.